File: Top 100 Codes/06-Advanced Codes related to Arrays/12-Count Inversion/CountInversion.py
def merge_and_count(arr, left, mid, right):
    # Create subarrays
    left_sub = arr[left:mid + 1]
    right_sub = arr[mid + 1:right + 1]
    
    i = 0  # index for left_sub
    j = 0  # index for right_sub
    k = left  # index for original array
    swaps = 0

    # Merge the two subarrays and count inversions
    while i < len(left_sub) and j < len(right_sub):
        if left_sub[i] <= right_sub[j]:
            arr[k] = left_sub[i]
            i += 1
        else:
            arr[k] = right_sub[j]
            j += 1
            swaps += len(left_sub) - i  # Count inversions
        k += 1

    # Copy remaining elements
    while i < len(left_sub):
        arr[k] = left_sub[i]
        i += 1
        k += 1

    while j < len(right_sub):
        arr[k] = right_sub[j]
        j += 1
        k += 1

    return swaps


def merge_sort_and_count(arr, left, right):
    count = 0
    if left < right:
        mid = (left + right) // 2

        count += merge_sort_and_count(arr, left, mid)
        count += merge_sort_and_count(arr, mid + 1, right)
        count += merge_and_count(arr, left, mid, right)

    return count


# Example usage
arr = [2, 4, 1, 3, 5]
result = merge_sort_and_count(arr, 0, len(arr) - 1)
print("Number of inversions:", result)

File: Top 100 Codes/06-Advanced Codes related to Arrays/12-Count Inversion/CountInversion.cpp
#include <iostream>
#include <vector>
using namespace std;

// Function to merge two subarrays and count inversions
int mergeAndCount(vector<int>& arr, int left, int mid, int right) {
    vector<int> leftSub(arr.begin() + left, arr.begin() + mid + 1);
    vector<int> rightSub(arr.begin() + mid + 1, arr.begin() + right + 1);

    int i = 0, j = 0, k = left, swaps = 0;

    // Merge while counting inversions
    while (i < leftSub.size() && j < rightSub.size()) {
        if (leftSub[i] <= rightSub[j]) {
            arr[k++] = leftSub[i++];
        } else {
            arr[k++] = rightSub[j++];
            swaps += (leftSub.size() - i); // Count the inversions
        }
    }

    // Copy remaining elements
    while (i < leftSub.size()) arr[k++] = leftSub[i++];
    while (j < rightSub.size()) arr[k++] = rightSub[j++];

    return swaps;
}

// Recursive merge sort function to count inversions
int mergeSortAndCount(vector<int>& arr, int left, int right) {
    int count = 0;
    if (left < right) {
        int mid = left + (right - left) / 2;

        count += mergeSortAndCount(arr, left, mid);
        count += mergeSortAndCount(arr, mid + 1, right);
        count += mergeAndCount(arr, left, mid, right);
    }
    return count;
}

int main() {
    vector<int> arr = {2, 4, 1, 3, 5};

    int inversionCount = mergeSortAndCount(arr, 0, arr.size() - 1);

    cout << "Number of inversions: " << inversionCount << endl;
    return 0;
}
